### PR TITLE
Add onempty handler for empty string, null and undefined

### PR DESCRIPTION
**Pull request by omrilotan**
Docs added:
```js
i18n.onempty((key, value, scope) => console.warn(`Empty value "${value}" for key "${key}" ${scope ? `In scope: "${scope}"`}`));
```

Allow custom fallbacks
```js
const i18n = new I18n({
    translations: {...},
    empty: (key, value, scope) => {
        if (scope.startsWith('en-US')) {
            return; // default fallback
        }

		log(`key ${key} has empty value ${value} in scope ${scope}`);
        return i18n.t(key, { $scope: 'en-US' }); // Try English
    },
    $scope: 'en-US'
});
```

### CODE REVIEW

File: tests/missing-key.js [hunk 1]
```diff
@@ -37,6 +37,28 @@ describe('missing keys report', () => {
         expect(reported).to.be.true;
     });
 
+    it('reports empty value for undefined', () => {
```
Comment (sahariko) [inline tests/missing-key.js]: I would add a similar test for `null` and empty strings as well.

File: index.js [hunk 2]
```diff
@@ -93,6 +98,12 @@ class I18n {
         // Handle one,other translation structure
         result = getOneOther(result, data);
 
+        if (EMPTY_VALUES.includes(result) && typeof this[EMPTY] === 'function') {
+            return this[EMPTY](
+                `${key}`, `${result}`, this.$scope, this.translations
```
Comment (sahariko) [inline index.js]:
```suggestion
                `${key}`, result, this.$scope, this.translations
```

I think stringifying the empty value might be a little confusing here.

File: index.js [hunk 3]
```diff
@@ -149,10 +164,30 @@ class I18n {
      * @return {self}
      *
      * @example
-     * i18n.onmiss((key) => logMissingKeyEvent({key: key.replace(/\W/g, '_'), })
+     * i18n.onmiss((key) => logMissingKeyEvent({key})
      */
     onmiss(callback) {
-        typeof callback === 'function' && this[MISSING].push(callback);
+        if (typeof callback === 'function') {
```
Comment (sahariko) [inline index.js:161]: You can afford to validate once (either here or on execution).

File: index.js (hunk 2 above)
Comment (omrilotan) [reply index.js]: Agreed

File: index.js (hunk 3 above)
Comment (omrilotan) [reply index.js:161]: Yes, implemented better

File: index.js [hunk 4]
```diff
@@ -22,12 +25,15 @@ const interpolate = paraphrase(/\${([^{}]*)}/g, /%{([^{}]*)}/g, /{{([^{}]*)}}/g)
  * @param {Object<Object>}   options.translations JSON compliant object
  * @param {Object<String>}   [options.$scope]     Root string to be use for looking for translation keys
  * @param {Object<Function>} [options.missing]    Method to call when key is not found
+ * @param {Object<Function>} [options.empty]      Method to call when value is empty
```
Comment (sahariko) [inline index.js]: [PETTY]

```suggestion
 * @param {Function} [options.empty]      Method to call when value is empty
```

Same for the other options.

File: index.js [hunk 5]
```diff
@@ -22,12 +25,15 @@ const interpolate = paraphrase(/\${([^{}]*)}/g, /%{([^{}]*)}/g, /{{([^{}]*)}}/g)
...
  */
 class I18n {
-    constructor({translations, $scope, missing} = {translations: {}, $scope: undefined, missing: undefined}) {
+    constructor({translations, $scope, missing, empty} = {translations: {}}) {
```
Comment (sahariko) [inline index.js]:
```suggestion
    constructor({translations, $scope, missing, empty} = {translations: {}, missing: () => undefined, empty: () => undefined}) {
```

**Review by sahariko — APPROVED** [review]

File: index.js (hunk 5 above)
Comment (omrilotan) [reply index.js]: Not exactly, you meant:
```suggestion
    constructor({translations, $scope, missing = () => null, empty = () => null} = {translations: {}}) {
```

But in actuality if someone passes a non function the while default system breaks. This way we're guaranteed onmiss and onempty are functions

Comment (sahariko) [reply index.js]: Yeah, my bad, meant to use it there.
